### PR TITLE
Handle errors during ansible-doc --metadata-dump more gracefully

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Right now ansible-doc dies if it encounters bad plugin/role metadata. That's not helpful for the VScode plugin or antsibull docs build.

This PR adjusts ansible-doc to catch these errors and return them in the plugin record with a key `error` (so this can be shown to the user if appropriate).

CC @bcoca @ssbarnea 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ansible-doc

### CODE REVIEW

**Comment by felixfontein** [comment]
Corresponding antsibull PR which uses this functionality: https://github.com/ansible-community/antsibull/pull/402

**Review by briantist — APPROVED** [review]
Looks like a decent solution for this, gives more leeway for callers to decide how to handle errors.

**Comment by bcoca** [comment]
I did something similar in my PR https://github.com/ansible/ansible/pull/74963/commits/8fbfae65bdbc2b99307fa8751a0a250717b22f31  but for testing we actually need them to 'break' since it prevents invalid docs from making it into the repo

**Comment by felixfontein** [comment]
@bcoca that's why the errors are only eaten for `--metadata-dump`, but not for regular operations :)

**Comment by bcoca** [comment]
@felixfontein  we specifically use metadata dump to fail plugins with incorrect docs, at best this would need to be a toggle

**Comment by felixfontein** [comment]
Ah, you mean in the integration tests. I was looking at the sanity tests, the ansible-doc test there already prevents broken plugins docs.

Do you want a toggle to explicitly enable failing (the test would have to be changed), or one to explicitly disable failing (antsibull and the VS code plugin would have to use it)?

**Review by bcoca — APPROVED** [review]

**Review by ssbarnea — APPROVED** [review]
I am i favour of this change but only have one minor thing to complain about `dont` is not a dictionary word and I would prefer better naming for the command line option, plain and simple like `--ignore-errors`.

Still, I give a +1 for the rest. Wait for another reviewer opinion before going for renaming the CLI, he might have another idea... ;)

**Comment by briantist** [comment]
> I am i favour of this change but only have one minor thing to complain about `dont` is not a dictionary word and I would prefer better naming for the command line option, plain and simple like `--ignore-errors`.

I agree that using a contraction like `dont` in a CLI switch like that is awkward, and I would avoid it or take the opportunity to change it if possible.

File: lib/ansible/cli/doc.py
Comment (cognifloyd) [inline lib/ansible/cli/doc.py]:
```suggestion
        self.parser.add_argument("--no-fail-on-errors", action="store_true", default=False, dest='dont_fail_on_errors',
```
A common pattern in CLI args is yup use "no".

**Comment by felixfontein** [comment]
I changed the option to @cognifloyd's suggestion. @ssbarnea @briantist does that sound good to you?

**Comment by mattclay** [comment]
@felixfontein If you need a file with a syntax error for use in an integration test, consider creating the file (or modifying it) at test time. We do that for other tests, such as those that require symbolic links.

**Comment by felixfontein** [comment]
@mattclay thanks for that suggestion!

File: lib/ansible/cli/doc.py
Comment (s-hertel) [inline lib/ansible/cli/doc.py:467]: This looks good to me. I don't think there's an easy way to add a suboption for one choice in a mutually exclusive group with argparse, but you could raise an AnsibleOptionsError in run() if this option is not provded with --metadata-dump.

Comment (felixfontein) [reply lib/ansible/cli/doc.py:467]: That's a great idea! I've implemented it in 5b0f970dc74777c8f43844d2312c75f644e0ae2d.

**Review by s-hertel — APPROVED** [review]

File: lib/ansible/cli/doc.py
Comment (bcoca) [inline lib/ansible/cli/doc.py]: don't do this, just make sure it is part of a metadata dump exclusive group when adding instead of making it part of the general group, this also avoids it appearing at all under the help of other options.

**Review by bcoca — DISMISSED** [review]

File: lib/ansible/cli/doc.py
Comment (s-hertel) [reply lib/ansible/cli/doc.py]: Assuming you mean
```diff
diff --git a/lib/ansible/cli/doc.py b/lib/ansible/cli/doc.py
index d5c57de00d..0e87699f04 100755
--- a/lib/ansible/cli/doc.py
+++ b/lib/ansible/cli/doc.py
@@ -463,7 +463,7 @@ class DocCLI(CLI, RoleMixin):
         exclusive.add_argument("--metadata-dump", action="store_true", default=False, dest='dump',
                                help='**For internal use only** Dump json metadata for all entries, ignores other options.')
 
-        self.parser.add_argument("--no-fail-on-errors", action="store_true", default=False, dest='no_fail_on_errors',
+        exclusive.add_argument("--no-fail-on-errors", action="store_true", default=False, dest='no_fail_on_errors',
                                  help='**For internal use only** Only used for --metadata-dump. '
                                       'Do not fail on errors. Report the error message in the JSON instead.')
```
, that would make it mutually exclusive with --metadata-dump. We'd need a conditional suboption for the exclusive group, but I don't know of a good way to do that with argparse. In ansible-galaxy we inspect the raw args, which is a hack, so I didn't suggest it here (but does allow showing the correct option help).

Comment (bcoca) [reply lib/ansible/cli/doc.py]: yes, i meant to create a option group (which has options exclusive to the top option, but I get the confusion)

Comment (felixfontein) [reply lib/ansible/cli/doc.py]: I reverted the last commit that added this check. Let's try to find a good solution for this in a follow-up PR instead of block this PR on this check.

Comment (s-hertel) [reply lib/ansible/cli/doc.py]: I would work on this with some guidance from @bcoca. It seems like there are some existing TODOs for ansible-doc options anyway, and learning how to fix this would probably help me clean up ansible-galaxy options too.

**Comment by felixfontein** [comment]
@s-hertel @bcoca @briantist @ssbarnea @cognifloyd @mattclay thanks for all your work, reviews, and comments on this!

**Pull request merged**
